### PR TITLE
Merging PVNet and XGboost

### PR DESCRIPTION
**Pull request by zakwatts**
# Pull Request

## Description

To merge together PVNet and XGboost OCF forecasts for the DRS backtest.
Scripts have been created to:
- Interpolate the XGb forecast to get 30 minute forecast durations.
- Merge and blend the 8 hours of PVNet with the rest of the time horizons between 0-40 being filled with XGb
- PVNet is just available for 2020-2023, so this has been taken account of.
- Pull PVNet forecast from Google
- Format an already normalised forecast
- Strip an existing forecast from previous probabilistic values.
- Script to un-normalise a forecast.

The repo itself has been reorganised.

## How Has This Been Tested?

Tests have been written for the the interpolation and merge and blend script.

Notebooks have also been created to aid in visualising the errors from the forecasts as well as to output the 0-8 hour and 0-40 hour error, as well as features to compare forecasts after a certain start date.

## Checklist:

- [ ] My code follows [OCF's coding style guidelines](https://github.com/openclimatefix/.github/blob/main/coding_style.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: scripts/analysis_forecasts_init_time.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import pandas as pd
```
Comment (peterdudfield) [inline scripts/analysis_forecasts_init_time.py]: could you add a sentence or two, of what this script does?

File: scripts/merge_scripts/format_forecast_no_norm.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+- 1 Hour Forecast
+....
+- 40 Hour Forecast
+
+The output csv file will have the following columns:
+- start_datetime_utc: datetime - the start datetime of the period
+- end_datetime_utc: datetime - the end datetime of the period
+- forecasting_creation_datetime_utc: datetime, when the forecast is made
+- generation_mw: float - the solar generation value
+"""
```
Comment (peterdudfield) [inline scripts/merge_scripts/format_forecast_no_norm.py:16]: nice

File: scripts/merge_scripts/gs_data_download.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
+"""
```
Comment (peterdudfield) [inline scripts/merge_scripts/gs_data_download.py:1]: nice

File: scripts/merge_scripts/merge_and_blend.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+- If blending is enabled, the script calculates blended forecasts for specific forecast horizons based on predefined 
+  blending ratios. These ratios determine the weight of each model's forecast in the final blended output.
+- For columns present in both datasets (excluding 'Init Time'), the script either blends the forecasts (if blending 
+  is enabled and the column is eligible for blending) or replaces XGBoost's forecasts with PVNet's forecasts.
+- The combined dataset is then filtered to include only data points starting from January 1, 2020, up to August 8, 2022.
+- A copy of the combined dataset is created to adjust the '0 Hour Forecast' by using PVNet's '0.5 Hour Forecast' for 
+  the next initiation time, compensating for XGBoost's lack of a '0 Hour Forecast'.
+- The script also prepares a separate dataset containing XGBoost's forecasts prior to January 1, 2020.
+- Finally, the pre-2020 XGBoost dataset and the adjusted combined dataset are concatenated to form the final merged 
+  and blended dataset, which is then returned.
```
Comment (peterdudfield) [inline scripts/merge_scripts/merge_and_blend.py:20]: nice nice nice

File: tests/test_merge_and_blend.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        ), f"Mismatch in '8.5 Hour Forecast' for Init Time {current_time}"
+
+
+# Test without blending
+def test_blend_data_without_blending(sample_data_xg, sample_data_pvnet):
+    result = blend_data(sample_data_xg, sample_data_pvnet, blend=False)
+    # Example check: ensure that columns from data_pvnet overwrite those in data_xg
+    assert result["7 Hour Forecast"][0] == sample_data_pvnet["7 Hour Forecast"][0]
+    assert result["8 Hour Forecast"][0] == sample_data_pvnet["8 Hour Forecast"][0]
+    assert result["9 Hour Forecast"][0] == sample_data_xg["9 Hour Forecast"][0]
```
Comment (peterdudfield) [inline tests/test_merge_and_blend.py:125]: nice two tests

**Review by peterdudfield — APPROVED** [review]

File: scripts/analysis_forecasts_init_time.py (hunk 1 above)
Comment (zakwatts) [reply scripts/analysis_forecasts_init_time.py]: This script is a duplicate of the missing_dates.py script so I have removed it.

**Pull request merged**